src/fluree/db/query/fql/parse.cljc

```diff
@@ -575,22 +575,22 @@
                 triples
                 (map vector (range) list))
 
+        ;; literal object
         (some? value)
-        (let [obj-cmp (if (v/variable? value)
+        (let [m*      (cond-> m language (assoc :lang language))
+              obj-cmp (if (v/variable? value)
                         (parse-variable value)
-                        (cond-> {::where/val value}
-                          (or m language) (assoc ::where/m (cond-> m language (assoc :lang language)))
-                          type (assoc ::where/datatype type)))]
+                        (where/anonymous-value value type m*))]
           (conj triples [subj-cmp pred-cmp obj-cmp]))
 
+        ;; ref object
         :else
-        (let [ref-cmp (if (nil? id)
-                        {::where/val (temp-bnode-id bnode-counter) ::where/datatype const/iri-id}
-                        (cond-> {::where/val id ::where/datatype const/iri-id}
-                          m (assoc ::where/m m)))
+        (let [ref-cmp (-> (where/match-iri (if (nil? id) (temp-bnode-id bnode-counter) id))
+                          (assoc ::where/datatype const/iri-id)
```

Comment:
The `::where/datatype` key should always be an internal numerical data type code. It's not set in the `match-iri` function because anything with a matched iri should *always* be an iri datatype.